src/main/scala/viper/silver/plugin/standard/reasoning/analysis/VarAnalysisGraphMap.scala

```diff
@@ -307,94 +405,102 @@ case class VarAnalysisGraphMap(prog: Program,
+      case _ => Set.empty
     }
   }
 
-  private def lookupVar(variable: Declaration, graphMap: GraphMap): Set[LocalVarDecl] = {
+  private def lookupVar(variable: InfluenceSink, graphMap: GraphMap): Set[InfluenceSource] = {
     // Assume Nodes are added when they are first encountered(can be a method call or a assume / inhale statement), so they might not exist when looked up
-    if(variable.isInstanceOf[AssumeNode]) {
-      graphMap.getOrElse(variable, Set())
-    } else {
-      assert(graphMap.contains(variable), "Variable " + variable + " not found in graph analysis")
-      graphMap(variable)
+    variable match {
+      case _: AssumeInfluenceSink => graphMap.getOrElse(variable, Set.empty)
+      case _ =>
+        assert(graphMap.contains(variable), "Variable " + variable + " not found in graph analysis")
+        graphMap(variable)
     }
   }
 
   /**
    * Gets all variables used in a given expressions and maps them to their influences specified in given graph map
    */
-  private def getResolvedVarsFromExp(exp: Exp, graphMap: GraphMap): Set[LocalVarDecl] = {
-    getLocalVarDeclsFromExpr(exp).flatMap(v => lookupVar(v, graphMap))
+  private def getResolvedVarsFromExp(exp: Exp, graphMap: GraphMap): Set[InfluenceSource] = {
+    getSinksFromExpr(exp).flatMap(v => lookupVar(v, graphMap))
   }
 
-  /**
-   * returns all the variables inside an expression
-   *
-   * @param exp expressions from which all variables should be returned
-   * @return set of Variable declarations
-   */
-  def getLocalVarDeclsFromExpr(exp: Exp): Set[LocalVarDecl] = {
+  /** returns all the sinks inside an expression */
+  def getSinksFromExpr(exp: Exp): Set[NonAssumeInfluenceSink] = {
     exp match {
-      case l@LocalVar(_, _) => Set(getLocalVarDeclFromLocalVar(l))
-      case BinExp(exp1, exp2) => getLocalVarDeclsFromExpr(exp1) ++ getLocalVarDeclsFromExpr(exp2)
-      case UnExp(exp) => getLocalVarDeclsFromExpr(exp)
+      case l: LocalVar => Set(LocalVarSink(l))
+      case BinExp(exp1, exp2) => getSinksFromExpr(exp1) ++ getSinksFromExpr(exp2)
+      case UnExp(exp) => getSinksFromExpr(exp)
       case f: FuncApp =>
-        val heapDependent = Expressions.couldBeHeapDependent(f.func(prog), prog)
-        f.args.flatMap(e => getLocalVarDeclsFromExpr(e).filter(v => v.typ != Ref)).toSet ++ (if (heapDependent) Set(heapVertex) else Set())
-      case DomainFuncApp(_, exps, _) => exps.flatMap(e => getLocalVarDeclsFromExpr(e)).toSet
-      case _: ForPerm | _: CurrentPerm => Set(heapVertex)
-      case FieldAccess(v, _) => getLocalVarDeclsFromExpr(v) + heapVertex
-      case AccessPredicate(access, _) => getLocalVarDeclsFromExpr(access) + heapVertex
-      case _ => Set()
+        val isHeapDependent = Expressions.couldBeHeapDependent(f.func(prog), prog)
+        // TODO: do we need to filter out Refs?
+        f.args.flatMap(e => getSinksFromExpr(e)).toSet ++ (if (isHeapDependent) Set(HeapSink) else Set.empty)
```

Comment:
Here I was unsure about porting the existing code. The `Ref` filtering imho seems to get rid of the heap node but wouldn't this get added anyway by taking the union if the singleton set containing the heap? I'm literally missing the intuition behind the existing code as it seems unclear why it's okay to filter out `Ref`

Reply:
You are right, I did not question this when I first saw it. I think your solution is right, as this example should not be allowed imo but would pass with the previous approach:
```
function test(e: Ref): Ref {
    e
}

method testM() {
    var x: Ref
    prove forall i:Ref assuming true implies i == x {
        x := test(i)
    }
}
```

Reply:
I agree. Please add this as a testcase